File: 2343-count-unguarded-cells-in-the-grid/2343-count-unguarded-cells-in-the-grid.cpp
class Solution {
public:
    int countUnguarded(int r, int c, vector<vector<int>>& guards, vector<vector<int>>& walls) {
        vector<vector<char>> grid(r, vector<char>(c, '0'));
        
        // Mark the positions of guards and walls on the grid
        for (const auto& guard : guards) {
            grid[guard[0]][guard[1]] = 'g';
        }
        for (const auto& wall : walls) {
            grid[wall[0]][wall[1]] = 'w';
        }
        
        int count = 0;  
        
        for (int i = 0; i < r; ++i) {
            // Right 
            int j = 0;
            while (j < c) {
                if (grid[i][j] == 'g') {
                    ++j;
                    while (j < c && grid[i][j] != 'w' && grid[i][j] != 'g') {
                        if (grid[i][j] == '1') {
                            ++j;
                            continue;
                        }
                        grid[i][j] = '1';
                        ++count;
                        ++j;
                    }
                } else {
                    ++j;
                }
            }
            
            // Left 
            j = c - 1;
            while (j >= 0) {
                if (grid[i][j] == 'g') {
                    --j;
                    while (j >= 0 && grid[i][j] != 'w' && grid[i][j] != 'g') {
                        if (grid[i][j] == '1') {
                            --j;
                            continue;
                        }
                        grid[i][j] = '1';
                        ++count;
                        --j;
                    }
                } else {
                    --j;
                }
            }
        }

        for (int j = 0; j < c; ++j) {
            // Down 
            int i = 0;
            while (i < r) {
                if (grid[i][j] == 'g') {
                    ++i;
                    while (i < r && grid[i][j] != 'w' && grid[i][j] != 'g') {
                        if (grid[i][j] == '1') {
                            ++i;
                            continue;
                        }
                        grid[i][j] = '1';
                        ++count;
                        ++i;
                    }
                } else {
                    ++i;
                }
            }

            // Up 
            i = r - 1;
            while (i >= 0) {
                if (grid[i][j] == 'g') {
                    --i;
                    while (i >= 0 && grid[i][j] != 'w' && grid[i][j] != 'g') {
                        if (grid[i][j] == '1') {
                            --i;
                            continue;
                        }
                        grid[i][j] = '1';
                        ++count;
                        --i;
                    }
                } else {
                    --i;
                }
            }
        }

        int res = (r * c) - (count + guards.size() + walls.size());
        
        return res;
    }
};

File: 2343-count-unguarded-cells-in-the-grid/2343-count-unguarded-cells-in-the-grid.py
class Solution:
    def countUnguarded(self, r: int, c: int, guards: List[List[int]], walls: List[List[int]]) -> int:
        grid = [[0] * c for _ in range(r)]
        
        for x, y in guards:
            grid[x][y] = "g"
        for x, y in walls:
            grid[x][y] = "w"
        
        count = 0  
        
        for i in range(r):
            # Right 
            j = 0
            while j < c:
                if grid[i][j] == "g":
                    j += 1
                    while j < c and grid[i][j] not in ("w", "g"):
                        if grid[i][j] == "1":
                            j += 1
                            continue
                        grid[i][j] = "1"
                        count += 1
                        j += 1
                else:
                    j += 1
            
            # Left 
            j = c - 1
            while j >= 0:
                if grid[i][j] == "g":
                    j -= 1
                    while j >= 0 and grid[i][j] not in ("w", "g"):
                        if grid[i][j] == "1":
                            j -= 1
                            continue
                        grid[i][j] = "1"
                        count += 1
                        j -= 1
                else:
                    j -= 1

        for j in range(c):
            # Down 
            i = 0
            while i < r:
                if grid[i][j] == "g":
                    i += 1
                    while i < r and grid[i][j] not in ("w", "g"):
                        if grid[i][j] == "1":
                            i += 1
                            continue
                        grid[i][j] = "1"
                        count += 1
                        i += 1
                else:
                    i += 1
            
            # Up 
            i = r - 1
            while i >= 0:
                if grid[i][j] == "g":
                    i -= 1
                    while i >= 0 and grid[i][j] not in ("w", "g"):
                        if grid[i][j] == "1":
                            i -= 1
                            continue
                        grid[i][j] = "1"
                        count += 1
                        i -= 1
                else:
                    i -= 1
        
        res = (r * c)- (count + len(guards) + len(walls))
        
        return res
